browser/resources/settings/brave_leo_assistant_page/model_config_ui.ts

```diff
@@ -0,0 +1,159 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+import 'chrome://resources/cr_elements/cr_button/cr_button.js'
+import 'chrome://resources/cr_elements/icons.html.js'
+
+import { PrefsMixin } from '/shared/settings/prefs/prefs_mixin.js'
+import { I18nMixin } from 'chrome://resources/cr_elements/i18n_mixin.js'
+import { PolymerElement } from 'chrome://resources/polymer/v3_0/polymer/polymer_bundled.min.js'
+import * as mojom from './brave_leo_assistant_browser_proxy.js'
+import { BaseMixin } from '../base_mixin.js'
+import { getTemplate } from './model_config_ui.html.js'
+
+const ModelConfigUIBase = PrefsMixin(I18nMixin(BaseMixin(PolymerElement)))
+
+class ModelConfigUI extends ModelConfigUIBase {
+  static get is() {
+    return 'model-config-ui'
+  }
+
+  static get template() {
+    return getTemplate()
+  }
+
+  static get properties() {
+    return {
+      label: {
+        type: String
+      },
+      modelRequestName: {
+        type: String
+      },
+      endpointUrl: {
+        type: String
+      },
+      apiKey: {
+        type: String,
+        value: ''
+      },
+      isUrlInvalid: {
+        type: Boolean,
+        value: false
+      },
+      modelItem: {
+        type: Object,
+        value: null,
+        notify: true,
+        observer: 'onModelItemChange_'
+      },
+      isEditing_: {
+        type: Boolean,
+        value: false,
+        readOnly: true,
+        computed: 'computeIsEditing_(modelItem)'
+      },
+      buttonLabel_: {
+        type: String,
+        computed: 'computeButtonLabel_(isEditing_)'
+      }
+    }
+  }
+
+  label: string
+  modelRequestName: string
+  endpointUrl: string
+  apiKey: string
+  modelItem: mojom.Model | null
+  isEditing_: boolean
+  isUrlInvalid: boolean
+
+  override ready() {
+    super.ready()
+  }
+
+  handleClick_() {
+    if (!this.saveEnabled_()) {
+      return
+    }
+
+    const mojomUrl = { url: '' }
+    mojomUrl.url = this.endpointUrl
+
+    // Adding random string to make sure the key is unique
+    // If editing, use the existing key
+    const modelKey = this.isEditing_
+      ? this.modelItem?.key
+      : `custom:${
+          new URL(this.endpointUrl).host
+        }:${this.modelRequestName.toLowerCase()}:${crypto
+          .randomUUID()
+          .slice(0, 8)}`
+
```

Comment:
Shouldn't we just use a uuid for the key - why do we need to involve the host? Wouldn't that fix any issues about model config changing (like host) and the Conversation's `model_key_` not referring to a valid Model anymore?

Reply:
initially, the UUID was derived from a combination of custom model options. however, i decided to add noise using randomUUID to avoid possible collisions and kept the host name to be friendly. this can help in scenario like determining which entry a model key might be related to. for example, ‘it seems like it’s the custom model and localhost entry.’

Reply:
i think this should be done in the service

Reply:
update: this has been moved to the service interface in cpp